### PR TITLE
Add JxW method to FEEvaluation

### LINKED ISSUE #3239: Add assert for FEEvaluation that reinit() is called prior to quadrature_point()
A rookie mistake, but I just noticed that if one tries to get a quadrature point from `FEEvaluation` without calliing `.reinit(cell)`, a segmentation fault occurs. I think `Assert()` is in order somewhere.

### LINKED ISSUE #3241: make FEEvaluation return JxW(q)
this is helpful when one needs to calculate integral quantities only in a cell / q-points loop.

### PR DESCRIPTION
**Pull request by kronbichler**
Fixes #3241.
Fixes #3239.

### CODE REVIEW

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 1]
```diff
@@ -2352,6 +2359,24 @@ ::fill_JxW_values(AlignedVector<VectorizedArray<Number> > &JxW_values) const
 
 
 
+template <int dim, int n_components_, typename Number>
+inline
+VectorizedArray<Number>
+FEEvaluationBase<dim,n_components_,Number>::JxW(const unsigned int q_point) const
+{
+  Assert (this->J_value != 0, ExcNotImplemented());
```
Comment (Rombur) [inline include/deal.II/matrix_free/fe_evaluation.h]: I think `ExcNotImplemented` is not the right assert here. You check that `J_value` has been initialized, right?

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 2]
```diff
@@ -2352,6 +2359,24 @@ ::fill_JxW_values(AlignedVector<VectorizedArray<Number> > &JxW_values) const
...
+  if (this->cell_type == internal::MatrixFreeFunctions::cartesian ||
+      this->cell_type == internal::MatrixFreeFunctions::affine)
+    {
+      Assert (this->mapping_info != 0, ExcNotImplemented());
```
Comment (Rombur) [inline include/deal.II/matrix_free/fe_evaluation.h]: Same here.

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 3]
```diff
@@ -6928,6 +6953,7 @@ ::quadrature_point (const unsigned int q) const
 {
   Assert (this->mapping_info->quadrature_points_initialized == true,
           ExcNotInitialized());
+  Assert (this->quadrature_pints != 0, ExcNotInitialized());
```
Comment (Rombur) [inline include/deal.II/matrix_free/fe_evaluation.h]: I think there is a typo it should be `quadrature_points`

File: tests/matrix_free/jxw_values.cc [hunk 4]
```diff
@@ -0,0 +1,110 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2013 - 2014 by the deal.II authors
```
Comment (Rombur) [inline tests/matrix_free/jxw_values.cc]: Copyright.

File: tests/matrix_free/jxw_values.cc [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+//
+// ---------------------------------------------------------------------
+
+
+
+// this function tests the correctness of the implementation of matrix free
+// operations in getting the function values, the function gradients, and the
+// function Laplacians on a mesh with all different cell types (same as in
+// matrix_vector_06) with hanging nodes, boundary conditions for float values.
+
```
Comment (Rombur) [inline tests/matrix_free/jxw_values.cc]: Is it really what you do?

Comment (kronbichler) [inline tests/matrix_free/jxw_values.cc]: No, this is a copy-paste leftover.

File: include/deal.II/matrix_free/fe_evaluation.h (hunk 3 above)
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation.h]: Right, I should have run all matrix-free tests before creating the PR...

File: include/deal.II/matrix_free/fe_evaluation.h (hunk 2 above)
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation.h]: This one should actually have been an internal error. You are right about the other ones.

**Comment by Rombur** [comment]
OK looks good.

**Comment by davydden** [comment]
hey, github again collapses outdated comments, hurray. :tada:

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 6]
```diff
@@ -2137,16 +2143,20 @@ ::FEEvaluationBase (const FEEvaluationBase<dim,n_components_,Number> &other)
...
-  jacobian           (other.jacobian),
-  J_value            (other.J_value),
-  quadrature_weights (other.quadrature_weights),
-  quadrature_points  (other.quadrature_points),
-  jacobian_grad      (other.jacobian_grad),
-  jacobian_grad_upper(other.jacobian_grad_upper),
-  cell               (other.cell),
-  cell_type          (other.cell_type),
-  cell_data_number   (other.cell_data_number),
+  cartesian_data     (0),
```
Comment (davydden) [inline include/deal.II/matrix_free/fe_evaluation.h]: is this change and below related to JxW? It looks like something else.

Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation.h]: This is related to the initialization check I added in `quadrature_point`: When you copy an object, you don't want it to be initialized to the state of the object copied from. In case the user unintentionally forgets to call `reinit()`, there will be some weird state and some pointers would be correctly set whereas others are invalid. I'd prefer to have invalid pointers so it gets noticed.

**Comment by kronbichler** [comment]
>look good, although i would split the changes to fe_evaluation.h into two commits, as they appear to be doing two different things.

I can split if you insist, but it was easier to verify all variants before committing. :)

**Comment by davydden** [comment]
Ah, ok then. Please merge yourself.
